### PR TITLE
bpo-35947: Update windows to the current version of libffi

### PR DESCRIPTION
**Pull request by paulmon**
This removes libffi_msvc and replaces it with the current version of libffi.
This builds on https://github.com/python/cpython/pull/3806
All test_ctypes tests pass on x86, and x64. Both debug and retail.
I also ran python -m test -j3 and there were no differences in test results before and after these changes.

https://bugs.python.org/issue35947

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by zooba** [comment]
Right now the build is failing on Windows because we haven't tagged the libffi sources. I'd rather not do that until we have it all working, so for now, can you change the `get-externals` references to simply `libffi` without the `3.3`? That should pull it from the branch tip, and simplify things if we need to patch it any more.

Once we're ready to merge, we can tag the libffi sources and update the reference here.

**Comment by paulmon** [comment]
There are some configured headers missing from cpython-source-deps
Also need a matching change python.props for the change in get_externals.bat

**Comment by paulmon** [comment]
The CI build is now blocked by a missing fficonfig.h.
I added this PR to cpython-source-deps to add the missing header files: https://github.com/python/cpython-source-deps/pull/7

**Comment by paulmon** [comment]
I moved the configured headers to modules/_ctypes/libffi_msvc so that they are not part of the libffi snapshot but generated from the snapshot, and manually edited.  I think this is all working now

File: PCbuild/_ctypes.vcxproj
Comment (zware) [inline PCbuild/_ctypes.vcxproj]: Looks like missing quotes at the ends of both of these lines.

File: PCbuild/_ctypes.vcxproj
Comment (zware) [inline PCbuild/_ctypes.vcxproj]: Same missing quotes.

File: Misc/NEWS.d/next/Windows/2019-02-11-14-53-17.bpo-35947.9vI4hP.rst
Comment (zware) [inline Misc/NEWS.d/next/Windows/2019-02-11-14-53-17.bpo-35947.9vI4hP.rst]: I would rather `libffi_msvc` go away entirely.  The collections in our `cpython-source-deps` repository contain changes like this already (or at least have historically).  We can tag the clean sources and the configured sources separately, but I'd rather keep all of the libffi code together and not in this repo :)

In fact, if we could make it possible to build libffi separately and just link _ctypes to it (analogous to `./configure --with-system-ffi` on UNIX systems), that would be ideal.  I won't push too hard for that right now, though; I'm just happy to finally be getting rid of this ancient copy of libffi :)

**Review by zware — CHANGES_REQUESTED** [review]
Just a few minor comments and I'm not qualified to review the ctypes change, otherwise :tada: :smile:

File: Misc/NEWS.d/next/Windows/2019-02-11-14-53-17.bpo-35947.9vI4hP.rst
Comment (paulmon) [reply Misc/NEWS.d/next/Windows/2019-02-11-14-53-17.bpo-35947.9vI4hP.rst]: I moved the configured files back to cpython-source-deps: https://github.com/python/cpython-source-deps/pull/8
I think pre-building a static lib or dll for libffi and putting in cpython-bin-deps should be a future work item. I would rather put effort towards adding arm64 support to libffi, after the arm32 changes are done.

File: PCbuild/_ctypes.vcxproj
Comment (paulmon) [reply PCbuild/_ctypes.vcxproj]: fixed

File: PCbuild/_ctypes.vcxproj
Comment (paulmon) [reply PCbuild/_ctypes.vcxproj]: fixed

**Comment by paulmon** [comment]
Moved configured files to cpython-source-deps as requested, and fixed quotes.
https://github.com/python/cpython-source-deps/pull/8 will need to be merged before CI will pass now
Re-ran all of the build and test_ctypes tests to confirm that things look good locally

**Comment by zooba** [comment]
> python/cpython-source-deps#8 will need to be merged before CI will pass now

It's merged. You can close/reopen to trigger CI again.

File: Misc/NEWS.d/next/Windows/2019-02-11-14-53-17.bpo-35947.9vI4hP.rst
Comment (zware) [reply Misc/NEWS.d/next/Windows/2019-02-11-14-53-17.bpo-35947.9vI4hP.rst]: :+1: sounds good to me.

File: PCbuild/_ctypes.vcxproj
Comment (zware) [reply PCbuild/_ctypes.vcxproj]: Still missing on the second line in both places.

Comment (paulmon) [reply PCbuild/_ctypes.vcxproj]: Quotes should be fixed now on both lines in both places.

File: PCbuild/_ctypes.vcxproj
Comment (zware) [inline PCbuild/_ctypes.vcxproj]: We still have `libffi_msvc` here, it should go away.

**Review by zware — CHANGES_REQUESTED** [review]
One final thing.

**Comment by zware** [comment]
@zooba Does `libffi-3.3.0-rc0-r1` sound like a reasonable tag for python/cpython-source-deps@8ba2b2c38744420a235e3e7f419cce2591aaf331?

File: PCbuild/_ctypes.vcxproj
Comment (paulmon) [reply PCbuild/_ctypes.vcxproj]: done

**Comment by zooba** [comment]
@zware Works for me. I'd also like to have the update process documented somewhere (possibly just as a `prepare_libffi.bat` file, if it's simpler). Hopefully we'll get to 3.3.0 before Python 3.8 releases.

**Comment by zware** [comment]
Ok, I've gone ahead and pushed the tag.

File: PCbuild/get_externals.bat
Comment (zware) [inline PCbuild/get_externals.bat]: The `libffi-3.3.0-rc0-r1` tag is now pushed:

```suggestion
set libraries=%libraries%                                       libffi-3.3.0-rc0-r1
```

File: PCbuild/python.props
Comment (zware) [inline PCbuild/python.props]:
```suggestion
    <libffiDir>$(ExternalsDir)libffi-3.3.0-rc0-r1\</libffiDir>
```

**Review by zware — COMMENTED** [review]
And now just an update to the new tag and this is good to go as far as I'm concerned.

Thank you for finally finishing this off!

File: PCbuild/python.props
Comment (paulmon) [reply PCbuild/python.props]: I'm happy to help.
Thanks for the feedback

**Review by zware — APPROVED** [review]

**Comment by paulmon** [comment]
> @zware Works for me. I'd also like to have the update process documented somewhere (possibly just as a `prepare_libffi.bat` file, if it's simpler). Hopefully we'll get to 3.3.0 before Python 3.8 releases.

The configured file was hand-edited to merge changes from the original PR and generated configured header.  I don't know how to document this.  

The sysv_intel.S and ffi.c changes for x86 changes were merged into libffi https://github.com/libffi/libffi/pull/468.  However I still need to figure out why autoconf is not working for msvc x86 when building libffi with cygwin. 

Once that works we could potentially check in configured files for each CPU type based on the output of the autoconf build step instead of having one merged fficonfig.h  

Does the documentation step need to be done before these changes can be merged?  Or can there be a seperate issue to follow up on?

**Comment by zooba** [comment]
@paulmon Let's put rough notes about the update process into the readme.txt file.

Do you have logs for the Cygwin failure? Technically that isn't required for us to merge this, but if we can figure it out relatively easily then I'd prefer to figure it out.

**Comment by paulmon** [comment]
Which readme.txt?  PCBuild/readme.txt or cpython-sources-dep/master/readme.rst? or some other file?
The steps for x64 were to follow the appveyor.yml steps from libffi manually.  I had to add x86 support and I tried to update the autoconf file but it still says the build type is not supported.  The PR for the Win32 changes was merged into master, and there is an open issue I created that I plan to follow up on: https://github.com/libffi/libffi/issues/469.  I can make time to look at that today, and update the readme

**Comment by paulmon** [comment]
The win32 libffi builds the way it's checked into libffi right now, I just had the command line parameters wrong.  The last commit adds a script that shows what I did to configure libffi based on the libffi/.appveyor.yml.

I cant' find the comment but I think it was suggested we that we switch to prebuilding libffi.dll and checking it in. The cygwin build scripts don't currently support building a .dll because msvcc.sh is missing support, so that's a bigger work item than just adding another build flag.

**Comment by paulmon** [comment]
prepare_libffi.bat still needs work to more fully automate updating libffi sources.
I wanted to save my progress in documenting this so far.

**Comment by paulmon** [comment]
updated prepare_libffi.bat to copy the configured files to include\<arch>\ffi*.h in the source tree to make it easier to check them in.  Configured files still require some patches for cpython.  I checked in the baseline and then patched in a different commit in this PR: https://github.com/python/cpython-source-deps/pull/10

**Comment by paulmon** [comment]
Added configure_headers.bat to cpython-source-deps/libffi
https://github.com/python/cpython-source-deps/pull/10

Updated cypthon-source-deps/master/README.rst
https://github.com/python/cpython-source-deps/pull/11

The headers only need to be configured not patched to build and pass all ctypes tests

**Comment by paulmon** [comment]
Unless I'm missing something this is ready to be merged.  What are the next steps?

**Comment by paulmon** [comment]
@zooba @zware 

While I was working on the libffi autoconf for I learned how to build the libffi-7.dll from libffi sources.
I just pushed the changes needed to use libffi-7.dll instead of including the libffi sources in _ctypes.
These changes pass all of the test_ctypes tests for x86/x64 both debug and release.

The changes to _ctypes.vcxproj will require pre-building libffi*.dll (see prepare_libffi.bat) and populating libffi-bin-3.3.0-rc0-r1 before they _ctypes will build,  which reminds me I need to update get_externals.bat to get the prebuilt libffi*.dll still.

Also the win32 changes got merged into libffi/master so you should be able to take a new snapshot and build libffi-bin-3.3.0-rc0-r1 from unpatched sources.  arm32 changes for libffi are in a PR that is still waiting for attention.

File: PCbuild/pyproject.props
Comment (paulmon) [inline PCbuild/pyproject.props]: After I installed VS 2019 to test something else this wouldn't build anymore without the %0D%0A after each line.  I reported the issue.  Let me know if you want to do something different here.

**Comment by paulmon** [comment]
If VS 2019 is installed then build.bat finds the new msbuild.exe which causes a build error.  Setting the MSBUILD environment variable back to the VS 2017 msbuild.exe is a better workaround than changing the project.   I have reported the build problem and am following up on it: https://developercommunity.visualstudio.com/content/problem/487337/writelinestofile-doesnt-write-new-lines.html

**Comment by zooba** [comment]
@paulmon I got a successful build of libffi - it's on cpython-bin-deps. Please switch your version name to just `libffi` for now and I'll tag it once we know the build works with the rest of your changes.

Make sure you sync your code too - I pushed some batch file changes to your branch.

**Comment by paulmon** [comment]
@zooba Looks good.  
After I sync'd I changed get_externals.bat to only get the cpython-bin-deps libffi by default. 
I also updated python.props.  
I built x86/x64, release and debug, and ran test_ctypes on all of them again.

**Comment by zooba** [comment]
@zware In case you want to review again, feel free (not sure exactly what's changed since your last one). Otherwise I'll try and get through this today and get it merged.

**Comment by zooba** [comment]
The Ubuntu failure on Azure Pipelines has been happening randomly all day. I wonder if a test was updated that's changing permissions on the temp folder? (They run in random order)

**Comment by zooba** [comment]
Like the change for PC/layout, there's a `tools/msi/lib` file (lib_files.wxs?) that will need a reference to the new DLL. It should already have the SSL references in it.

Otherwise, I think the rest looks fine.

**Review by zooba — APPROVED** [review]

**Comment by zooba** [comment]
The Pipelines failure is due to an unrelated Linux issue, so ignoring that check to merge.

**Comment by zware** [comment]
Thanks again @paulmon for seeing this through!  It's been a goal for a few years now, it's nice to see it finally happen :)

**Pull request merged**
